adrs/1358452048/README.md

```diff
@@ -0,0 +1,118 @@
+# Slots reserved React property key
+
+* **State:** Draft
+* **Created:** 2020-12-14
+* **Tags:** react
+
+## Context
+
+You may have components that you would like the users to have control over
+some underline components.
+
+For example, let's assume that we would like to change the container and the
+label components for the following component.
+
+```tsx
+function Badge(props) {
+  const displayValue = '...' // calculate something, does not matter the details
+
+  return (
+    <div className="container">
+      {props.children}
+      <div className="label">
+        {displayValue}
+      </div>
+    </div>
+  )
+}
+```
+
+Also, although some of the implementations are not technically wrong, they may
+miss some concern.
+
+For example, using render function over components:
+
+```tsx
+function Badge(props) {
+  return (
+    <div className="container">
+      {props.children}
+      {props.renderLabel({ className: 'label', children: displayValue })}
+    </div>
+  )
+}
+```
+
+In the previous example, React loses the ability to figure out if the component
+needs to be re-render since it is no longer a component but a function call.
+
+Also, subjectively speaking, it feels non-React since we lose the JSX syntax.
+
+We have no alignment in how to accomplish the tasks, which means people have to
+learn new idioms across codebase even when the intention is the same.
+
+What key should we use for this? Should we pass a function or a component, or
+something else? Defining the TypeScript definitions and so on.
+
+## Resolution
```

Comment:
OK! What do you think about having a recursive structure for slots?

Let me explain:

The point of Slots is to increase flexibility to help authors create reusable components. 


Some of the prior art of this work is the theme-ability of component styles.  I posted in slack the library I used to theme components for a couple of years: https://github.com/javivelasco/react-css-themr.  There are similar libraries that gained much more traction, but the idea is the same.  These libraries aimed to solve a problem that I believe this ADR has; the ability to customize components should be as composable as the components themselves. 

In other words, the usefulness of Slots starts breaking down when you start composing components, especially if you only have access to the first layer of a component's slots.

In the case of composing themes, this library would concat component names as namespaces.  So if an `Input` component has a `label` CSS class, and a `Dropdown` component uses the `Input` component, you could pass in a theme with the class `DropdownInputLabel` that would be threaded down through `Dropdown` -> `Input` -> `Label`.



I came up with an example to illustrate this point:


```tsx
import * as React from 'react';

// You could reuse this across your system
type PropsWithSlots<
  C extends Partial<{ [key: string]: React.ElementType }>,
  P
> = P & {
  slots?: C;
};

type FancyInputSlots = Partial<{
  Input: React.ElementType<
    Pick<JSX.IntrinsicElements['input'], 'type' | 'value'>
  >;
  Label: React.ElementType<Pick<JSX.IntrinsicElements['div'], 'className'>>;
}>;

type FancyInputProps = PropsWithSlots<
  FancyInputSlots,
  {
    value: string;
  }
>;

const DefaultLabel = 'div';

function FancyInput(props: FancyInputProps) {
  const Input = props.slots?.Input ?? 'input';
  const Label = props.slots?.Label ?? DefaultLabel;

  return (
    <Label className="label">
      <Input type="text" value={props.value} />
    </Label>
  );
}

type ReusableFormSlots = Partial<{
  Root: React.ElementType<Pick<JSX.IntrinsicElements['div'], 'children'>>;
  Username: React.ElementType<FancyInputProps>;
  Password: React.ElementType<FancyInputProps>;
}>;

type ReusableFormProps = PropsWithSlots<ReusableFormSlots, {}>;

function ReusableForm(props: ReusableFormProps) {
  const Root = props.slots?.Root ?? 'div';
  const Username = props.slots?.Username ?? FancyInput;
  const Password = props.slots?.Password ?? FancyInput;

  const values = { username: 'Joe Sixpack', password: 'hunter42' };
  return (
    <Root>
      <Username value={values.username} />
      <Password value={values.password} />
    </Root>
  );
}

// ================================================================

// IMPLEMENTATION

// CASE 1

function AllDefault() {
  return <ReusableForm />;
}

// CASE 2

function PasswordInput(props: JSX.IntrinsicElements['input']) {
  return <input {...props} type="password" />;
}
function CustomPassword(props: FancyInputProps) {
  const slots = { ...props.slots, Input: PasswordInput };
  return <FancyInput {...props} slots={slots} />;
}

function OverwriteSlots() {
  const slots: ReusableFormProps['slots'] = { Password: CustomPassword };
  return <ReusableForm slots={slots} />;
}

// CASE 3

function OverwriteNestedSlots() {
  const slots: ReusableFormProps['slots'] = {
    Password: { Input: PasswordInput },
  };
  return <ReusableForm slots={slots} />;
}

```

Reply:
I am confused about the "recursive structure for slots".

From my perspective, if you need to customize underline components, use the key, but the ADR is not touching in people abusing or creating messy APIs around component designs, for example, don't allow people to compose the components, and force them to use `slots`.

But I am not sure what do  you want me to answer 😄